stor/s3.py

```diff
@@ -41,21 +41,25 @@ def _parse_s3_error(exc, **kwargs):
     http_status = exc.response.get('ResponseMetadata', {}).get('HTTPStatusCode')
     msg = exc.response['Error'].get('Message', 'Unknown')
     code = exc.response['Error'].get('Code')
-
-    # Give some more info about the error's origins
-    if 'Bucket' in kwargs:
-        msg += ' Bucket: ' + kwargs.get('Bucket')
-    if 'Key' in kwargs:
-        msg += ', Key: ' + kwargs.get('Key')
+    operation_name = exc.operation_name
+    msg += ' Bucket: ' + kwargs.get('Bucket') if 'Bucket' in kwargs else ''
+    msg += ' Key: ' + kwargs.get('Key') if 'Key' in kwargs else ''
 
     if http_status == 403:
         if 'storage class' in msg and code == 'InvalidObjectState':
-            return exceptions.ObjectInColdStorageError(msg, exc)
+            if operation_name == 'GetObject':
+                return exceptions.ObjectInColdStorageError(msg, exc)
+            if operation_name == 'RestoreObject':
+                return exceptions.AlreadyRestoredError(msg, exc)
         return exceptions.UnauthorizedError(msg, exc)
-    elif http_status == 404:
+    if http_status == 404:
         return exceptions.NotFoundError(msg, exc)
-    elif http_status == 503:
+    if http_status == 503:
         return exceptions.UnavailableError(msg, exc)
+    if http_status == 409:
+        if 'Object restore is already in progress' in msg:
```

Comment:
I guess we just have to hope amazon keeps message the same. Or does this come from `boto`?

Reply:
again, kinda can take it as it comes... if amazon updates its API, we'll update this library :-/